File: hub/src/light/effect/effect_combiner.py
from light.effect.resolved_effect import ResolvedEffect
from light.rgbw import RGBW

class EffectCombiner:
    def __init__(self, effects):
        self.unresolvedEffects = effects

    def combine(self, dt):
        color = self.combineColors(dt)
        amplitude = self.combineAmplitude(dt)
        ledPositions = self.combinePositions(dt)
        return ResolvedEffect(color, amplitude, ledPositions)


    def combineColors(self, dt):
        combinedColor = RGBW(0, 0, 0, 0)
        for unresolvedEffect in self.unresolvedEffects:
            resolvedEffect = unresolvedEffect.getEffect(dt)
            if (resolvedEffect.rgbw):
                combinedColor.add(resolvedEffect.rgbw)
        return combinedColor

    def combineAmplitude(self, dt):
        amplitude = 0
        for unresolvedEffect in self.unresolvedEffects:
            resolvedEffect = unresolvedEffect.getEffect(dt)
            if (resolvedEffect.amplitude):
                amplitude = amplitude + resolvedEffect.amplitude
        return amplitude

    def combinePositions(self, dt):
        positions = []
        for unresolvedEffect in self.unresolvedEffects:
            resolvedEffect = unresolvedEffect.getEffect(dt)
            print(resolvedEffect.ledPositions)
            if (resolvedEffect.ledPositions):
                positions.extend(resolvedEffect.ledPositions)
        return list(set(positions))

File: hub/drumlite-js/src/effect/EffectCombiner.ts
import PartialEffect from "../light/effect/PartialEffect";
import ResolvedEffect from "./ResolvedEffect";
import RGB from "../light/RGB";

export default class EffectCombiner {
    unresolvedEffects: PartialEffect<any>[];
    
    constructor(effects: PartialEffect<any>[]) {
        this.unresolvedEffects = effects;
    }

    public combine(dt: number): ResolvedEffect[] {
        const resolvedEffects: ResolvedEffect[] = [];
        let color = new RGB();
        let amplitude = 0;
        let ledPositions: number[] = [];
        this.unresolvedEffects.forEach(unresolvedEffectGroup => {
            unresolvedEffectGroup.getEffect(dt).forEach(resolvedEffect => {
                color = this.combineColors(resolvedEffect, dt, color);
                amplitude = this.combineAmplitude(resolvedEffect, dt, amplitude);
                ledPositions = this.combinePositions(resolvedEffect, dt, ledPositions);
            });
        });
        resolvedEffects.push(new ResolvedEffect(color, amplitude, ledPositions));
        if (this.unresolvedEffects.find(e => e.isJit())) {
            console.log({resolved: resolvedEffects});
        }
        return resolvedEffects;
    }

    public combineColors(effectGroup: ResolvedEffect[], dt: number, color: RGB) {
        for (let effect of effectGroup) {
            if (effect.rgb) {
                color.add(effect.rgb);
            }
        }
        return color;
    }

    public combineAmplitude(effectGroup: ResolvedEffect[], dt: number, amplitude: number) {
        for (let effect of effectGroup) {
            if (effect.amplitude) {
                amplitude = amplitude + effect.amplitude;
            }
        }
        return amplitude;
    }

    public combinePositions(effectGroup: ResolvedEffect[], dt: number, positions: number[]): number[] {
        for (let effect of effectGroup) {
            if (effect.ledPositions) {
                positions.push(...effect.ledPositions);
            }
        }
        return [...new Set<number>(positions).keys()];
    }
}
